velox/exec/tests/DriverTest.cpp

```diff
@@ -818,6 +826,294 @@ TEST_F(DriverTest, driverCreationThrow) {
+              ASSERT_EQ(
+                  driver->task()->enterSuspended(driver->state()),
+                  testData.expectedEnterSuspensionStopReason);
+              suspensionNotify.notify();
+              leaveSuspensionWait.wait(leaveSuspensionKey);
+              ASSERT_EQ(
+                  driver->task()->leaveSuspended(driver->state()),
+                  testData.expectedLeaveSuspensionStopReason);
+              leaveSuspensionNotify.notify();
+            })));
+
+    task->start(task, testData.numDrivers, 1);
+    driverExecutionWait.wait(driverExecutionWaitKey);
+
+    if (testData.enterSuspensionAfterTaskTerminated) {
+      task->requestCancel();
+      enterSuspensionWait.notify();
+      suspensionNotify.wait(suspensionNotifyKey);
+    } else {
+      enterSuspensionWait.notify();
+      suspensionNotify.wait(suspensionNotifyKey);
+      task->requestCancel();
+    }
+    leaveSuspensionWait.notify();
+    leaveSuspensionNotify.wait(leaveSuspensionNotifyKey);
+    waitForTaskAborted(task.get(), 1000'000'000);
+  }
+}
+
+DEBUG_ONLY_TEST_F(DriverTest, driverSuspensionCalledFromOffThread) {
+  auto vector = makeRowVector({"c0"}, {makeFlatVector<int32_t>({1, 2, 3})});
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  auto plan =
+      PlanBuilder(planNodeIdGenerator).values({vector}, true).planFragment();
+  auto task = std::make_shared<exec::Task>(
+      "t0",
+      plan,
+      0,
+      std::make_shared<core::QueryCtx>(driverExecutor_.get()),
+      [](RowVectorPtr /*unused*/, ContinueFuture* /*unused*/) {
+        return exec::BlockingReason::kNotBlocked;
+      });
+
+  std::shared_ptr<Driver> driver;
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::exec::Values::getOutput",
+      std::function<void(const exec::Values*)>([&](const exec::Values* values) {
+        driver = values->testingOperatorCtx()->driver()->shared_from_this();
+      }));
+  task->start(task, 1, 1);
+  ASSERT_TRUE(waitForTaskCompletion(task.get(), 100'000'000));
+  while (driver->isOnThread()) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+  }
+  ASSERT_ANY_THROW(driver->task()->enterSuspended(driver->state()));
+  ASSERT_ANY_THROW(driver->task()->leaveSuspended(driver->state()));
+}
+
+DEBUG_ONLY_TEST_F(DriverTest, driverSuspensionLeaveBlockedByPendingPause) {
+  auto vector = makeRowVector({"c0"}, {makeFlatVector<int32_t>({1, 2, 3})});
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  auto plan =
+      PlanBuilder(planNodeIdGenerator).values({vector}, true).planFragment();
+  auto task = std::make_shared<exec::Task>(
+      "t0",
+      plan,
+      0,
+      std::make_shared<core::QueryCtx>(driverExecutor_.get()),
+      [](RowVectorPtr /*unused*/, ContinueFuture* /*unused*/) {
+        return exec::BlockingReason::kNotBlocked;
+      });
+
+  std::atomic<bool> injectSuspension{true};
+  folly::EventCount driverExecutionWait;
```

Comment:
These counters seem to be reused in multiple tests. Would it make sense to create a helper struct for them to make the code a bit easier to read?

Reply:
Yeah, thought about this but the race conditions in each test has some subtle difference and the triggering involves the test value. So it is a bit hard to reuse in code.

Reply:
Ok. I found a way to reuse them by adding a test function. Thanks for the suggestion :)